core/common/src/main/java/alluxio/resource/CloseableIterator.java

```diff
@@ -18,32 +18,67 @@
 import java.io.IOException;
 import java.util.Iterator;
 import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.function.Consumer;
 
 /**
  * A {@code CloseableIterator<T>} is an iterator which requires cleanup when it is no longer in use.
  *
+ * When the user creates {@code CloseableIterator}, the {@code closeResource()} method needs to be
+ * implemented to define how the resource can be closed.
+ * The iterator is not necessarily {@code Closeable} so we do not provide a default implementation
+ * for {@code closeResource()}.
+ *
  * @param <T> the type of elements returned by the iterator
  */
-public abstract class CloseableIterator<T> extends CloseableResource<Iterator<T>> {
+public abstract class CloseableIterator<T> extends CloseableResource<Iterator<T>> implements Iterator<T> {
+  Iterator<T> mIter;
+
   /**
    * Creates a {@link CloseableIterator} wrapper around the given iterator. This iterator will
    * be returned by the {@link CloseableIterator#get()} method.
    *
    * @param iterator the resource to wrap
    */
-  public CloseableIterator(Iterator<T> iterator) {
+  CloseableIterator(Iterator<T> iterator) {
     super(iterator);
+    mIter = iterator;
+  }
+
+  Iterator<T> getIterator() {
+    return mIter;
+  }
+
+  @Override
+  public boolean hasNext() {
+    return mIter.hasNext();
+  }
+
+  @Override
+  public T next() {
+    return mIter.next();
   }
 
+  public static <T> CloseableIterator<T> create(Iterator<T> iterator, Consumer<Void> closeAction) {
+    return new CloseableIterator<T>(iterator) {
+      @Override
+      public void closeResource() {
+        closeAction.accept(null);
```

Comment:
The close action here depends purely on outside variables and does not have a return value. There's no functional interface that takes no argument and returns nothing (well, there's `Runnable` but I don't think that's a good use). Here I preferred `Consumer` because it only relies on side effects. But this may not be the best option.